### PR TITLE
Optimization method FormsService::canSubmit

### PR DESCRIPTION
**Pull request by AIlkiv**
Currently, to check if a user has filled out the form, all entities are retrieved, and then two loops are performed at the PHP. My version does the same thing but with a single simple query.

### CODE REVIEW

File: lib/Db/SubmissionMapper.php [hunk 1]
```diff
@@ -87,26 +87,27 @@ public function findById(int $id): Submission {
 
 	/**
 	 * @param int $formId
-	 * @throws DoesNotExistException if not found
-	 * @return array
+	 * @param string $userId
+	 * @return bool
 	 */
-	public function findParticipantsByForm(int $formId): array {
+	public function hasSubmissions(int $formId, string $userId): bool {
```
Comment (Chartman123) [inline lib/Db/SubmissionMapper.php:93]:
```suggestion
	public function hasFormSubmissionsByUser(int $formId, string $userId): bool {
```

What about including form and user in the function name? You could also please add some description of what it does to the PHPDoc comment.

**Review by Chartman123 — COMMENTED** [review]
Nice one :) Just a little comment about the new name for the function :)

@susnux what do you think?

And it would be nice if you could add some more tests to satisfy the Codecov checks 😉

File: lib/Db/SubmissionMapper.php (hunk 1 above)
Comment (AIlkiv) [reply lib/Db/SubmissionMapper.php]: @Chartman123 As for the tests, I'm not sure if it makes sense to write tests for DB Mapper, since other classes of database work are not covered.

File: lib/Db/SubmissionMapper.php [hunk 2]
```diff
@@ -86,44 +86,48 @@ public function findById(int $id): Submission {
...
+	 * @param int $formId ID of the form
+	 * @param string $userId UID of the user
+	 * @param bool $checkMultipleFormSubmissions if true, we check if there is more than one submissions, if false, at least one submissions
+	 * @return bool
 	 */
-	public function findParticipantsByForm(int $formId): array {
+	public function hasFormSubmissionsByUser(
+		int $formId,
+		string $userId,
+		bool $checkMultipleFormSubmissions
```
Comment (susnux) [inline lib/Db/SubmissionMapper.php:98]: Hm maybe make this numeric and directly use for the `requireCountFormSubmissions` ?
If not I would make this false by default.

File: lib/Db/SubmissionMapper.php [hunk 3]
```diff
@@ -86,44 +86,48 @@ public function findById(int $id): Submission {
...
+	): bool {
+		$requireCountFormSubmissions = $checkMultipleFormSubmissions ? 2 : 1;
+
 		$qb = $this->db->getQueryBuilder();
 
-		$qb->select('user_id')
+		$query = $qb->select('id')
```
Comment (susnux) [inline lib/Db/SubmissionMapper.php:104]: Maybe just:
```suggestion
		$query = $qb->select($qb->createFunction('COUNT(`id`)'))
```

File: lib/Db/SubmissionMapper.php [hunk 4]
```diff
@@ -86,44 +86,48 @@ public function findById(int $id): Submission {
...
 	/**
-	 * Count submissions by form and optionally also by userId
-	 * @param int $formId ID of the form to count submissions for
-	 * @param string|null $userId optionally limit submissions to the one of that user
+	 * Count submissions by form
+	 * @param int $formId ID of the form to count submissions
 	 * @throws \Exception
 	 */
-	public function countSubmissions(int $formId, ?string $userId = null): int {
+	public function countSubmissions(int $formId): int {
```
Comment (susnux) [inline lib/Db/SubmissionMapper.php:125]: This is not related

**Review by susnux — COMMENTED** [review]
Generally nice! Some notes :)

(and we should add a composite index on "formId userId"

File: lib/Db/SubmissionMapper.php (hunk 4 above)
Comment (AIlkiv) [reply lib/Db/SubmissionMapper.php:111]: countSubmissions with userId only used on isUniqueSubmission. isUniqueSubmission is removed in this PR.

File: lib/Db/SubmissionMapper.php [hunk 5]
```diff
@@ -86,36 +86,39 @@ public function findById(int $id): Submission {
...
 	/**
-	 * Count submissions by form and optionally also by userId
-	 * @param int $formId ID of the form to count submissions for
+	 * Count submissions by form with optional filters
+	 * @param int $formId ID of the form to count submissions
 	 * @param string|null $userId optionally limit submissions to the one of that user
 	 * @throws \Exception
 	 */
-	public function countSubmissions(int $formId, ?string $userId = null): int {
+	protected function countSubmissionsWithFilters(int $formId, ?string $userId = null, int $limit = -1): int {
```
Comment (Chartman123) [inline lib/Db/SubmissionMapper.php:121]: `@param int $limit` is missing in PHPDoc comment

Comment (AIlkiv) [reply lib/Db/SubmissionMapper.php:122]: fix

File: lib/Db/SubmissionMapper.php (hunk 3 above)
Comment (AIlkiv) [reply lib/Db/SubmissionMapper.php]: fixed

File: lib/Db/SubmissionMapper.php (hunk 2 above)
Comment (AIlkiv) [reply lib/Db/SubmissionMapper.php]: I created two separate methods. I think this is more correct.

**Comment by Chartman123** [comment]
@AIlkiv could you please rebase on current main?

File: lib/Controller/ApiController.php [hunk 6]
```diff
@@ -1161,7 +1161,8 @@ public function insertSubmission(int $formId, array $answers, string $shareHash
 
 		// Ensure the form is unique if needed.
 		// If we can not submit anymore then the submission must be unique
-		if (!$this->formsService->canSubmit($form) && !$this->submissionService->isUniqueSubmission($submission)) {
+		if (!$this->formsService->canSubmit($form) && $this->submissionMapper->hasMultipleFormSubmissionsByUser($form, $submission->getUserId())
+		) {
```
Comment (Chartman123) [inline lib/Controller/ApiController.php:1165]: I'd like to keep the braces in the first line like everywhere else :)

```suggestion
		if (!$this->formsService->canSubmit($form) && $this->submissionMapper->hasMultipleFormSubmissionsByUser($form, $submission->getUserId())) {
```

File: lib/Db/SubmissionMapper.php [hunk 7]
```diff
@@ -124,14 +128,16 @@ public function countSubmissions(int $formId, ?string $userId = null): int {
...
+			$query->setMaxResults($limit);
+		}
 
 		$result = $query->executeQuery();
 		$row = $result->fetch();
 		$result->closeCursor();
 
 		return (int) ($row['num_submissions'] ?? 0);
 	}
-
```
Comment (Chartman123) [inline lib/Db/SubmissionMapper.php:134]: Please add this empty line again :)

File: lib/Migration/Version040300Date20240708163401.php [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	 * @param array $options
+	 * @return null|ISchemaWrapper
+	 */
+	public function changeSchema(IOutput $output, Closure $schemaClosure, array $options): ?ISchemaWrapper {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+
+		if ($schema->hasTable('forms_v2_submissions')) {
+			$schema->getTable('forms_v2_submissions')
+				->addIndex(['form_id', 'user_id'], 'forms_submissions_form_user');
```
Comment (Chartman123) [inline lib/Migration/Version040300Date20240708163401.php:47]: Please check if the index isn't already there before adding it

File: lib/Service/FormsService.php [hunk 9]
```diff
@@ -333,15 +333,12 @@ public function canSubmit(Form $form): bool {
 		if ($this->currentUser->getUID() === $form->getOwnerId()) {
 			return true;
 		}
-
```
Comment (Chartman123) [inline lib/Service/FormsService.php:336]: Please add empty line again

File: lib/Service/FormsService.php [hunk 10]
```diff
@@ -333,15 +333,12 @@ public function canSubmit(Form $form): bool {
...
-		if (!$form->getSubmitMultiple()) {
-			$participants = $this->submissionMapper->findParticipantsByForm($form->getId());
-			foreach ($participants as $participant) {
-				if ($participant === $this->currentUser->getUID()) {
-					return false;
-				}
-			}
+		if (
+			!$form->getSubmitMultiple()
+			&& $this->submissionMapper->hasFormSubmissionsByUser($form, $this->currentUser->getUID())
```
Comment (Chartman123) [inline lib/Service/FormsService.php:339]: I'd put the `&&` in the first line

```suggestion
			!$form->getSubmitMultiple() &&
            $this->submissionMapper->hasFormSubmissionsByUser($form, $this->currentUser->getUID())
```

**Review by Chartman123 — COMMENTED** [review]
Some small comments, but rest seems ok to me :)

File: lib/Migration/Version040300Date20240708163401.php (hunk 8 above)
Comment (AIlkiv) [reply lib/Migration/Version040300Date20240708163401.php:47]: I checked on migrations and in my local database.

Comment (Chartman123) [reply lib/Migration/Version040300Date20240708163401.php:47]: Yes but I mean that you should add an `if` for it and only create it when it's not already there. :)

Comment (AIlkiv) [reply lib/Migration/Version040300Date20240708163401.php]: fixed

**Review by Chartman123 — APPROVED** [review]
